resolver/dns/dns_resolver_test.go

```diff
@@ -1126,3 +1142,80 @@ func TestCustomAuthority(t *testing.T) {
 		}
 	}
 }
+
+// TestRateLimitedResolve exercises the rate limit enforced on re-resolution
+// requests. It sets the re-resolution rate to a small value and repeatedly
+// calls ResolveNow() and ensures only the expected number of resolution
+// requests are made.
+func TestRateLimitedResolve(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	dc := replaceDNSResRate(250 * time.Millisecond)
+	defer dc()
+
+	// Create a new testResolver{} for this test because we want the exact count
+	// of the number of times the resolver was invoked.
+	nc := replaceNetFunc()
+	defer nc()
+
+	target := "foo.bar.com"
+	b := NewBuilder()
+	cc := &testClientConn{target: target}
+	r, err := b.Build(resolver.Target{Endpoint: target}, cc, resolver.BuildOption{})
+	if err != nil {
+		t.Fatalf("resolver.Build() returned error: %v\n", err)
+	}
+	defer r.Close()
+
+	dnsR, ok := r.(*dnsResolver)
+	if !ok {
+		t.Fatalf("resolver.Build() returned unexpected type: %T\n", dnsR)
+	}
+	tr, ok := dnsR.resolver.(*testResolver)
+	if !ok {
+		t.Fatalf("delegate resolver returned unexpected type: %T\n", tr)
+	}
+
+	// We set the re-resolution rate to 250ms at the start of this test. We
+	// expect the first call to ResolveNow() to make a resolution request
+	// immediately (T0), and one amongst the other queued up calls to make
+	// another resolution request at (T0+250ms). We wait till a little longer to
+	// make sure that more resolution requests are not made.
+	start := time.Now()
+	end := start.Add(100 * time.Millisecond)
+	calls := 0
+	for time.Now().Before(end) {
+		r.ResolveNow(resolver.ResolveNowOption{})
+		calls++
+		time.Sleep(1 * time.Millisecond)
+	}
+	if calls <= 2 {
+		t.Fatalf("expected to make more than two calls to ResolveNow, made only %d", calls)
+	}
+	time.Sleep(1 * time.Second)
+
+	wantAddrs := []resolver.Address{{Addr: "1.2.3.4" + colonDefaultPort}, {Addr: "5.6.7.8" + colonDefaultPort}}
+	var gotAddrs []resolver.Address
+	var cnt int
+	for {
+		gotAddrs, cnt = cc.getAddress()
+		if cnt > 0 {
+			break
+		}
+		time.Sleep(time.Millisecond)
+	}
+	if !reflect.DeepEqual(gotAddrs, wantAddrs) {
+		t.Errorf("Resolved addresses of target: %q = %+v, want %+v\n", target, gotAddrs, wantAddrs)
+	}
+
+	// The actual number of resolution requests made could be 2 or 3, because
+	// when the resolver is built in Build(), we also create a timer initialized
+	// to time.Duration(0) and which is eventually reset to defaultFreq (30 mins)
```

Comment:
I don't understand this.  Something sounds wrong.  I don't see where `Build` in the DNS (name) path of the dnsResolver will do a name resolution automatically.  And even then, it should be predictable what happens.

Reply:
I was referring to this in Build() where we initialize a timer:
```go
	d := &dnsResolver{
		freq:                 b.minFreq,
		backoff:              backoff.Exponential{MaxDelay: b.minFreq},
		host:                 host,
		port:                 port,
		ctx:                  ctx,
		cancel:               cancel,
		cc:                   cc,
		t:                    time.NewTimer(0),
		rn:                   make(chan struct{}, 1),
		disableServiceConfig: opts.DisableServiceConfig,
	}
```

And in `watcher()` where we select on a bunch of things:
```go
func (d *dnsResolver) watcher() {
	defer d.wg.Done()
	for {
		select {
		case <-d.ctx.Done():
			return
		case <-d.t.C:
		case <-d.rn:
		}
```

And for the first iteration of this loop, sometimes we read from d.t.C and sometimes from d.rn and that changes the number of resolution requests (2 or 3, respectively).

Reply:
Aha.  In that case, let's wait until the first resolution happens before starting the `ResolveNow` storm - we should be able to do that with signals instead of sleeps, though.

Reply:
Actually - I think the problem is when we reset the timer - we should do a `Stop()` and drain the channel before resetting, in case the timer elapses at the same time a `ResolveNow` call comes in.

https://golang.org/pkg/time/#Timer.Reset

Reply:
Done.
But I also needed a `select` while attempting to drain the channel. Is that expected?

Reply:
No, that doesn't sound right.  As the documentation of `Reset` says, if `t.Stop()` returns `false` you should be able to unconditionally drain the timer channel.  (Unless it's being `select`ed on concurrently in another goroutine.)

Reply:
Moving the `t.Stop()` call as mentioned in the other comment seems to have helped. Thanks.